### PR TITLE
Fix #1675 - ForwardTranslate WaterHeater:Mixed when not on loop but has Peak Use Flow Rate (standalone operation)

### LINKED ISSUE #1675: WaterHeater:Mixed not forward translated when not on loop
If a WaterHeater:Mixed is not on a SWH PlantLoop then it is not forward translated. 

A water heater can be standalone when PeakUseFlowRate and UseFlowRateFractionSchedule are set.

I came across this in a few prototype buildings where this is used.

This iddobjectype just need to be added to [here](https://github.com/NREL/OpenStudio/blob/8e67be624fef6e4182f66df4782f45195422cf25/openstudiocore/src/energyplus/ForwardTranslator.cpp#L1997)

One option around this is to create a plantloop with a pump that has basically zero head pressure and a water use equipment that is the same as the PeakUseFlowRate and has a schedule the same as UseFlowRateFractionSchedule.

fyi @kbenne

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

- Fixes #1675
- ForwardTranslate WaterHeater:Mixed when not on loop but has Peak Use Flow Rate (standalone operation)

Please read [OpenStudio Pull Requests](https://github.com/NREL/OpenStudio/wiki/OpenStudio-Pull-Requests) to better understand the OpenStudio Pull Request protocol.

### Pull Request Author

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Checked behavior in OpenStudioApplication, adjusted policies as needed (`src/openstudio_lib/library/OpenStudioPolicy.xml`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/energyplus/ForwardTranslator.cpp [hunk 1]
```diff
@@ -3600,6 +3600,12 @@ std::vector<IddObjectType> ForwardTranslator::iddObjectsToTranslateInitializer()
   result.push_back(IddObjectType::OS_Coil_Heating_Water);
   result.push_back(IddObjectType::OS_Coil_Heating_WaterToAirHeatPump_EquationFit);
   result.push_back(IddObjectType::OS_Coil_WaterHeating_Desuperheater);
+
+  // If using a plantLoop, this will be translated by the PlantLoop. But WaterHeaters can also be used stand-alone, so always translate them
+  // We'll check in their FT if the "Peak Use Flow Rate" is actually initialized as it's an indication that the WH was
+  result.push_back(OS_WaterHeater_Mixed);
+  result.push_back(OS_WaterHeater_Stratified);
```
Comment (jmarrec) [inline src/energyplus/ForwardTranslator.cpp:3607]: Always translate them...

File: src/energyplus/ForwardTranslator/ForwardTranslateWaterHeaterMixed.cpp [hunk 2]
```diff
@@ -58,6 +58,12 @@ boost::optional<IdfObject> ForwardTranslator::translateWaterHeaterMixed( WaterHe
   boost::optional<double> value;
   boost::optional<Schedule> schedule;
 
+  if (!modelObject.plantLoop() && !(modelObject.peakUseFlowRate() && modelObject.useFlowRateFractionSchedule()) ) {
+    LOG(Warn, modelObject.briefDescription() << " will not be translated as it not on a PlantLoop, and it does not have both a Peak Use Flow Rate "
+        "and a Use Flow Rate Fraction Schedule which is a required condition for stand-alone operation");
+    return boost::none;
+  }
```
Comment (jmarrec) [inline src/energyplus/ForwardTranslator/ForwardTranslateWaterHeaterMixed.cpp:65]: BUT do not actually translate them to IDF if either:
* On a PlantLoop
* has both the peakUseFlowRate set (optional double) **and the** useFlowRateFractionSchedule (optional schedule).


Technically I think we could do without the useFlowRateFractionSchedule but whatever, **it is desired here I think**, as it'll avoid people setting the flow rate by mistake (in OS App for eg...) and ending up with "hidden" standalone heaters if they get orphaned from a plantloop.

E+ source shows that no severe will be issued if schedule not specified while standalone operation, and it basically amounts to using an Unity schedule:

https://github.com/NREL/EnergyPlus/blob/68c51b98f26f703a0f7b92ac7c30a18f8a526313/src/EnergyPlus/WaterThermalTanks.cc#L2577-L2586

https://github.com/NREL/EnergyPlus/blob/68c51b98f26f703a0f7b92ac7c30a18f8a526313/src/EnergyPlus/WaterThermalTanks.cc#L5969-L5987


@kbenne

**Comment by kbenne** [comment]
@jmarrec I'm seeing this from the CI

```/Users/jenkins/git/OpenStudioIncremental/develop/src/energyplus/Test/WaterHeaterMixed_GTest.cpp:64:19: note: in instantiation of template class 'boost::optional<openstudio::model::Schedule>' requested here
  EXPECT_FALSE(wh.useFlowRateFractionSchedule());
```

Which I think normally indicates a missing `#include <Foo_Impl>` Schedule_Impl in this case probably. Can you take a look?

**Comment by jmarrec** [comment]
> @jmarrec I'm seeing this from the CI
> 
> ```
> /Users/jenkins/git/OpenStudioIncremental/develop/src/energyplus/Test/WaterHeaterMixed_GTest.cpp:64:19: note: in instantiation of template class 'boost::optional<openstudio::model::Schedule>' requested here
>   EXPECT_FALSE(wh.useFlowRateFractionSchedule());
> ```
> 
> Which I think normally indicates a missing `#include <Foo_Impl>` Schedule_Impl in this case probably. Can you take a look?

Thanks @kbenne I think I just forgot to include `Schedule.hpp`. I pushed a fix, we'll see what CI reports.

**Comment by jmarrec** [comment]
@kbenne I have a failing test and I **cannot** understand why.

From the Gtest itself, `waterHeater.plantLoop()` is perfectly `true` as it should. Yet within the FT, it is false?!?

https://github.com/NREL/OpenStudio/blob/8bcd27d41d882ce09357f90a193a63f443268067/src/energyplus/ForwardTranslator/ForwardTranslateWaterHeaterMixed.cpp#L66

```
[ RUN      ] EnergyPlusFixture.ForwardTranslatorGeneratorMicroTurbine_ELCD_Orphan
[BOOST_ASSERT] <2> Assertion idf_component failed on line 106 of boost::optional<openstudio::IdfObject> openstudio::energyplus::ForwardTranslator::translatePlantEquipmentOperationHeatingLoad(openstudio::model::PlantEquipmentOperationHeatingLoad&) in file /home/julien/Software/Others/OpenStudio/src/energyplus/ForwardTranslator/ForwardTranslatePlantEquipmentOperationHeatingLoad.cpp.
openstudio_energyplus_tests: /home/julien/Software/Others/OpenStudio/src/energyplus/Test/../../model/../utilities/core/Assert.hpp:61: void boost::assertion_failed(const char*, const char*, const char*, long int): Assertion `false' failed.
[1/1] EnergyPlusFixture.ForwardTranslatorGeneratorMicroTurbine_ELCD_Orphan returned/aborted with exit code -6 (402 ms)
FAILED TESTS (1/1):
     402 ms: Products/openstudio_energyplus_tests EnergyPlusFixture.ForwardTranslatorGeneratorMicroTurbine_ELCD_Orphan
Built branch '1675_WaterHeater_Standalone' in 1min7s, tests ran in 0min6s

```

**Comment by jmarrec** [comment]
Linux and Mac are happy, the ECLD_Orphan test now passes with the fix. Thanks @kbenne for helping me find the needle in the haystack.

This is ready to drop

**Pull request merged**